nemo_aligner/algorithms/ppo.py

```diff
@@ -101,86 +190,66 @@ def __init__(
-
-            # compute metrics
-            # NOTE: this metric is not accumulated globally so it will differ between DP ranks
-            ppo_rollout_metrics["init_policy_kl"] += (
-                masked_mean(init_policy_kl, mask, dim=-1).sum().item() if self.compute_init_policy_kl else 0
+        ppo_rollout_data = {}
+        ppo_rollout_metrics = {}
+
+        prompt_lengths = rollout_batch["prompt_lengths"]
+        response_lengths = rollout_batch["response_lengths"]
+        response_tokens = rollout_batch["response_tokens"]
+        values = rollout_batch["values"]
+        rewards = rollout_batch["rewards"]
+        logprobs = rollout_batch["logprobs"]
+        is_end = rollout_batch["is_end"]
+
+        if self.compute_init_policy_kl:
+            init_policy_kl = calculate_kl_penalty(
+                log_probs_a=rollout_batch["logprobs"],
+                log_probs_b=rollout_batch["init_logprobs"],
+                use_absolute_kl=self.cfg.use_absolute_kl,
             )
+        else:
+            init_policy_kl = torch.tensor(0, dtype=logprobs.dtype, device=logprobs.device)
 
-        # average across the samples for the non global metrics
-        ppo_rollout_metrics = {k: v / num_samples for k, v in ppo_rollout_metrics.items()}
+        rewards_with_kl = calculate_ppo_rewards(
+            values, rewards, response_lengths, init_policy_kl, self.cfg.initial_policy_kl_penalty
+        )
 
-        for k in ppo_rollout_data:
-            rollout_batch_seq_length = self.rollout_batch_seq_length
-            pad_value = self.model.tokenizer.eos_id
+        mask = create_mask(values=values, prompt_lengths=prompt_lengths, response_lengths=response_lengths)
+        advantages, returns = calculate_advantages_and_returns(
+            values=values,
+            rewards=rewards_with_kl,
+            discount_factor=self.cfg.discount_factor,
+            gae_lambda=self.cfg.gae_lambda,
+            mask=mask,
+        )
 
-            # all other tensors in the rollout batch
-            # will be B x S -1 (because we don't predict anything for the last token)
-            if k != "response_tokens":
-                pad_value = 0
-                if rollout_batch_seq_length is not None:
-                    rollout_batch_seq_length -= 1
+        # collect everything we need to train PPO
+        ppo_rollout_data["mask"] = mask
+        ppo_rollout_data["advantages"] = advantages
+        ppo_rollout_data["prev_logprobs"] = logprobs
+        ppo_rollout_data["response_tokens"] = response_tokens
+        ppo_rollout_data["is_end"] = is_end
+        # for the critic
+        ppo_rollout_data["values"] = values
+        ppo_rollout_data["returns"] = returns
+
+        # compute metrics
+        # these are not global yet
+        ppo_rollout_metrics["init_policy_kl"] = (
+            masked_mean(init_policy_kl, mask, dim=-1).sum().item() if self.compute_init_policy_kl else 0
```

Comment:
ah... good point. I went down the rabbit hole of masked tensors and found this to be a pretty neat solution: 9bfc3645fd4161bbf3b5b42405a45088d25b0413

it does raise some depreciation warnings though

Reply:
Ah, interesting!

1. Can you suppress this warning somehow with a context manager? (it seems to be quite spammy, at least in my console)
2. This is actually changing the output since `NaNs` become `0's` (kind of related to my other comment on `ppo_ratio`). Is that what we want? (`NaNs` can be annoying but they are also easier to notice in general, while zeros can pollute more silently).

Reply:
1. 8d915f30cbd4de5a4386e5d34fbc06f2ca30e75d -> I didn't find it too spammy but given we already have a million lines of printout this is a step in the right direction :)
2. 22bbdf9ace31187b10ba2902e32d1df3c384be9c -> good point, let's keep this as nan to indicate something went wrong